src/bridge/IrcBridge.ts

```diff
@@ -0,0 +1,1199 @@
+                labels: ["outcome"],
+            })
+        };
+
+        // Custom IRC metrics
+        const reconnQueue = metrics.addGauge({
+            name: "clientpool_reconnect_queue",
+            help: "Number of disconnected irc connections waiting to reconnect.",
+            labels: ["server"]
+        });
+
+        const memberListLeaveQueue = metrics.addGauge({
+            name: "user_leave_queue",
+            help: "Number of leave requests queued up for virtual users on the bridge.",
+            labels: ["server"]
+        });
+
+        const memberListJoinQueue = metrics.addGauge({
+            name: "user_join_queue",
+            help: "Number of join requests queued up for virtual users on the bridge.",
+            labels: ["server"]
+        });
+
+        const ircHandlerCalls = metrics.addCounter({
+            name: "irchandler_calls",
+            help: "Track calls made to the IRC Handler",
+            labels: ["method"]
+        });
+
+        const matrixHandlerConnFailureKicks = metrics.addCounter({
+            name: "matrixhandler_connection_failure_kicks",
+            help: "Track IRC connection failures resulting in kicks",
+            labels: ["server"]
+        });
+
+        metrics.addCollector(() => {
+            this.ircServers.forEach((server) => {
+                reconnQueue.set({server: server.domain},
+                    this.clientPool.totalReconnectsWaiting(server.domain)
+                );
+                const mxMetrics = this.matrixHandler.getMetrics(server.domain);
+                matrixHandlerConnFailureKicks.inc(
+                    {server: server.domain},
+                    mxMetrics["connection_failure_kicks"] || 0
+                );
+            });
+
+            Object.keys(this.memberListSyncers).forEach((server) => {
+                memberListLeaveQueue.set(
+                    {server},
+                    this.memberListSyncers[server].getUsersWaitingToLeave()
+                );
+                memberListJoinQueue.set(
+                    {server},
+                    this.memberListSyncers[server].getUsersWaitingToJoin()
+                );
+            });
+
+            const ircMetrics = this.ircHandler.getMetrics();
+            Object.keys(ircMetrics).forEach((method) => {
+                const value = ircMetrics[method];
+                ircHandlerCalls.inc({method}, value);
+            });
+        });
+    }
+
+    public getAppServiceUserId() {
+        return this.appServiceUserId as string;
+    }
+
+    public getStore() {
+        return this.dataStore;
+    }
+
+    public getAppServiceBridge() {
+        return this.bridge;
+    }
+
+    public getClientPool() {
+        return this.clientPool;
+    }
+
+    public getProvisioner() {
+        return this.provisioner;
+    }
+
+    public createBridgedClient(ircClientConfig: IrcClientConfig, matrixUser: MatrixUser|null, isBot: boolean) {
```

Comment:
re `matrixUser` being nullable, does this need to be a special case, maybe with another create* method? Seems strange to sometimes pass null here considering the whole goal is to get it all typed

Reply:
Yeah, I'm starting to think we should have a dedicated creator for bot clients.

Reply:
Logged as #852 because we should merge this first.